Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp

```diff
@@ -13319,20 +13319,42 @@ void SpeculativeJIT::compileStringReplace(Node* node)
         && node->child1().useKind() == StringUse
         && node->child2().useKind() == StringUse
         && node->child3().useKind() == StringUse) {
-        if (JSString* replace = node->child3()->dynamicCastConstant<JSString*>(); replace && !replace->length()) {
-            SpeculateCellOperand string(this, node->child1());
-            SpeculateCellOperand search(this, node->child2());
-            GPRReg stringGPR = string.gpr();
-            GPRReg searchGPR = search.gpr();
-            speculateString(node->child1(), stringGPR);
-            speculateString(node->child2(), searchGPR);
+        String replacement = node->child3()->tryGetString(m_graph);
+        if (!!replacement) {
+            if (!replacement.length()) {
+                SpeculateCellOperand string(this, node->child1());
+                SpeculateCellOperand search(this, node->child2());
+                GPRReg stringGPR = string.gpr();
+                GPRReg searchGPR = search.gpr();
+                speculateString(node->child1(), stringGPR);
+                speculateString(node->child2(), searchGPR);
 
-            flushRegisters();
-            GPRFlushedCallResult result(this);
-            callOperation(operationStringReplaceStringEmptyString, result.gpr(), JITCompiler::LinkableConstant(m_jit, m_graph.globalObjectFor(node->origin.semantic)), stringGPR, searchGPR);
-            m_jit.exceptionCheck();
-            cellResult(result.gpr(), node);
-            return;
+                flushRegisters();
+                GPRFlushedCallResult result(this);
+                callOperation(operationStringReplaceStringEmptyString, result.gpr(), JITCompiler::LinkableConstant(m_jit, m_graph.globalObjectFor(node->origin.semantic)), stringGPR, searchGPR);
+                m_jit.exceptionCheck();
+                cellResult(result.gpr(), node);
+                return;
+            }
+
+            if (replacement.find('$') == notFound) {
```

Comment:
Nit: This should really be in strength reduction

Reply:
Yeah, the plan is introducing `StringReplaceString` node which does more strength reduction (in a subsequent patch) in the right phase instead of here.